File: src/commands/fs.ts
import { type DevFsOp, devFs } from "../client.ts";
import { requireCredentials } from "../config.ts";

async function readStdin(): Promise<string> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) chunks.push(chunk as Buffer);
  return Buffer.concat(chunks).toString("utf8");
}

/**
 * The file-edit verbs (read/write/str-replace/ls/grep/rm). Each maps to one
 * `dev/fs` op against the experiment's live dev working tree. All require an
 * open dev node (`orx dev open`); the API returns a clear error otherwise.
 */
export async function fsCommand(
  verb: string,
  expId: string | undefined,
  rest: string[],
): Promise<void> {
  if (!expId) {
    console.error(`Usage: orx ${verb} <experimentId> ...`);
    process.exit(1);
  }
  const creds = await requireCredentials();

  let op: DevFsOp;
  switch (verb) {
    case "read":
      if (!rest[0]) return usage(`orx read <experimentId> <path>`);
      op = { op: "read", path: rest[0] };
      break;
    case "write":
      if (!rest[0]) return usage(`orx write <experimentId> <path>   (content on stdin)`);
      op = { op: "write", path: rest[0], content: await readStdin() };
      break;
    case "str-replace":
      if (rest.length < 3)
        return usage(`orx str-replace <experimentId> <path> <old_string> <new_string>`);
      op = { op: "str_replace", path: rest[0]!, old_string: rest[1]!, new_string: rest[2]! };
      break;
    case "ls":
      op = rest[0] ? { op: "list", path: rest[0] } : { op: "list" };
      break;
    case "grep":
      if (!rest[0]) return usage(`orx grep <experimentId> <pattern>`);
      op = { op: "search", query: rest[0] };
      break;
    case "rm":
      if (!rest[0]) return usage(`orx rm <experimentId> <path>`);
      op = { op: "delete", path: rest[0] };
      break;
    default:
      return usage(`unknown file command: ${verb}`);
  }

  const { output } = await devFs(creds, expId, op);
  console.log(output);
}

function usage(msg: string): never {
  console.error(`Usage: ${msg}`);
  process.exit(1);
}

File: src/commands/fs.rs
//! The file-edit verbs (read/write/str-replace/ls/grep/rm).
//!
//! Each maps to one `dev/fs` op against the experiment's live dev working tree.
//! All require an open dev node (`orx dev open`); the API returns a clear error
//! otherwise.

use tokio::io::AsyncReadExt;

use crate::client::{dev_fs, DevFsOp};
use crate::error::{require_credentials, Result};
use crate::{FsInvocation, FsVerb};

async fn read_stdin() -> Result<String> {
    let mut buf = String::new();
    tokio::io::stdin().read_to_string(&mut buf).await?;
    Ok(buf)
}

pub async fn run(inv: FsInvocation) -> Result<()> {
    let FsInvocation { verb, exp_id, rest } = inv;
    let creds = require_credentials().await;

    let op: DevFsOp = match verb {
        FsVerb::Read => DevFsOp::Read {
            path: rest[0].clone(),
        },
        FsVerb::Write => DevFsOp::Write {
            path: rest[0].clone(),
            content: read_stdin().await?,
        },
        FsVerb::StrReplace => DevFsOp::StrReplace {
            path: rest[0].clone(),
            old_string: rest[1].clone(),
            new_string: rest[2].clone(),
        },
        FsVerb::Ls => DevFsOp::List {
            path: rest.first().cloned(),
        },
        FsVerb::Grep => DevFsOp::Search {
            query: rest[0].clone(),
        },
        FsVerb::Rm => DevFsOp::Delete {
            path: rest[0].clone(),
        },
    };

    let output = dev_fs(&creds, &exp_id, &op).await?.output;
    println!("{}", output);
    Ok(())
}
